packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -0,0 +1,660 @@
+    removeVM,
+    getRenderRoot,
+    LWCDOMMode,
+} from './vm';
+import {
+    VNode,
+    VCustomElement,
+    VElement,
+    VNodes,
+    Key,
+    VText,
+    VComment,
+    VBaseElement,
+    VNodeType,
+} from './vnode';
+
+import { applyEventListeners } from './modules/events';
+import { patchAttributes } from './modules/attrs';
+import { patchProps } from './modules/props';
+import { patchClassAttribute } from './modules/computed-class-attr';
+import { patchStyleAttribute } from './modules/computed-style-attr';
+import { applyStaticClassAttribute } from './modules/static-class-attr';
+import { applyStaticStyleAttribute } from './modules/static-style-attr';
+
+interface KeyToIndexMap {
+    [key: string]: number;
+}
+
+const SVG_NAMESPACE = 'http://www.w3.org/2000/svg';
+
+export function patchChildren(parentElm: ParentNode, oldCh: VNodes, newCh: VNodes): void {
+    if (hasDynamicChildren(newCh)) {
+        updateDynamicChildren(parentElm, oldCh, newCh);
+    } else {
+        updateStaticChildren(parentElm, oldCh, newCh);
+    }
+}
+
+function patch(n1: VNode | null, n2: VNode, parent: ParentNode, anchor: Node | null) {
+    if (n1 === n2) {
+        return;
+    }
+
+    // FIXME: When does this occurs, is it possible with LWC?
+    if (!isNull(n1) && !sameVnode(n1, n2)) {
+        anchor = n2.owner.renderer.nextSibling(n1.elm);
+        unmount(n1, parent);
+        n1 = null;
+    }
+
+    switch (n2.type) {
+        case VNodeType.Text:
+            processText(n1 as VText, n2, parent, anchor);
+            break;
+
+        case VNodeType.Comment:
+            processComment(n1 as VComment, n2, parent, anchor);
+            break;
+
+        case VNodeType.Element:
+            processElement(n1 as VElement, n2, parent, anchor);
+            break;
+
+        case VNodeType.CustomElement:
+            processCustomElement(n1 as VCustomElement, n2, parent, anchor);
+            break;
+    }
+}
+
+function processText(n1: VText, n2: VText, parent: ParentNode, anchor: Node | null) {
+    const { owner } = n2;
+    const { renderer } = owner;
+
+    if (isNull(n1)) {
+        const textNode = (n2.elm = renderer.createText(n2.text));
+        linkNodeToShadow(textNode, owner);
+
+        insertVNode(n2, parent, anchor);
+    } else {
+        n2.elm = n1.elm;
+
+        if (n2.text !== n1.text) {
+            updateTextContent(n2);
+        }
+    }
+}
+
+function processComment(n1: VComment, n2: VComment, parent: ParentNode, anchor: Node | null) {
+    const { owner } = n2;
+    const { renderer } = owner;
+
+    if (isNull(n1)) {
+        const textNode = (n2.elm = renderer.createComment(n2.text));
+        linkNodeToShadow(textNode, owner);
+
+        insertVNode(n2, parent, anchor);
+    } else {
+        n2.elm = n1.elm;
+
+        // FIXME: Comment nodes should be static, we shouldn't need to diff them together. However
+        // it is the case today.
+        if (n2.text !== n1.text) {
+            updateTextContent(n2);
```

Comment:
I like the abstraction. don't know if will help but here is super micro-optimization: we could pass `(node, renderer, text)` and avoid destructing the vnode in `updateTextContent`. we have all those here.